### PR TITLE
Mark filtered check tests DISABLED instead of using return().

### PR DESCRIPTION
**Pull request by ScottTodd**
The filtering behavior in `iree_check_test` currently has early `return()` calls that skip defining tests entirely. With this PR, filtered tests are now defined but have the [`DISABLED`](https://cmake.org/cmake/help/latest/prop_test/DISABLED.html) test property set. We can build on this to improve how the project handles hardware-dependent tests: https://github.com/openxla/iree/issues/12042.

## What these changes do

Output from `ctest` clearly indicates disabled tests, showing which tests _could have been run_ if the implicit or explicit CMake configuration was different:

```
λ cmake -G Ninja -B ../iree-build-rt/ . -DIREE_BUILD_COMPILER=OFF
λ cmake --build ../iree-build-rt/
λ ctest --test-dir ../iree-build-rt/

Internal ctest changing into directory: D:/dev/projects/iree-build-rt
Test project D:/dev/projects/iree-build-rt
...
100% tests passed, 0 tests failed out of 146

Label Time Summary:
driver=cuda                                                   =  12.94 sec*proc (164 tests)
driver=local-sync                                             =   1.77 sec*proc (59 tests)
driver=local-task                                             =   2.67 sec*proc (351 tests)
driver=metal                                                  =   0.00 sec*proc (100 tests)
driver=vulkan                                                 =  14.05 sec*proc (138 tests)
...
Total Test time (real) =  48.79 sec

The following tests did not run:
          2 - iree/tests/e2e/linalg/check_llvm-cpu_local-task_conv2d.mlir (Disabled)
          3 - iree/tests/e2e/linalg/check_llvm-cpu_local-task_f32_to_i4.mlir (Disabled)
          4 - iree/tests/e2e/linalg/check_llvm-cpu_local-task_i4_to_f32.mlir (Disabled)
...
```

IDEs also show "disabled" tests differently: 
image

Sample CI logs (note that `ctest_all.sh` adds its own layer of filtering with `--label-exclude`):

Job | Notes | Logs link
-- | -- | --
`build_test_runtime` | Skip all 'check' tests | [Logs](https://github.com/openxla/iree/actions/runs/7468071370/job/20322826358?pr=16086#step:5:1211)
`test_all` | CUDA and Metal filtered by labels | [Logs](https://github.com/openxla/iree/actions/runs/7468240936/job/20323443107?pr=16086#step:6:2987)
`test_gpu` | No tests skipped, labels include only GPU tests | [Logs](https://github.com/openxla/iree/actions/runs/7468071370/job/20323028256?pr=16086#step:7:769)
`cross_compile_and_test (riscv_64)` | Skip `llvm-cpu-host` tests | [Logs](https://github.com/openxla/iree/actions/runs/7468071370/job/20323029021?pr=16086#step:8:1293)

## Implementation details

Check tests are composed of two main targets - `iree_bytecode_module()` and `iree_native_test()`. We used to skip both if any filter conditions matched. Now, there is more nuance. To make the filtering logic easier to follow, I've introduced some new local variables.

I also moved the target and driver filtering from `iree_check_single_backend_test_suite` to `iree_check_test`. This may increase CMake configure time since several values are recomputed for the ~1000 check tests. If we add more complicated logic (e.g. asking `iree-compile` in `IREE_HOST_BIN_DIR` which targets it supports), we should add some variable caching.

### CODE REVIEW

**Comment by ScottTodd** [comment]
CI builds seems fine to me, but I might have missed something and accidently enabled or disabled something unexpected. We don't (can't) have exhaustive tests for enabling and disabling subsets of the project. However, I think this does get us in a better place for spotting configuration bugs if something did slip through.

File: build_tools/cmake/test_riscv.sh [hunk 1]
```diff
@@ -34,6 +34,7 @@ ctest_args=(
 declare -a label_exclude_args=(
   "^nodocker$"
   "^driver=vulkan$"
+  "^driver=metal$"
   "^driver=cuda$"
```
Comment (ScottTodd) [inline build_tools/cmake/test_riscv.sh:38]: I'd like to eventually remove these label exclusions as part of https://github.com/openxla/iree/issues/12042. Keeping them consistent here for now.

Comment (ScottTodd) [reply build_tools/cmake/test_riscv.sh:38]: (so users not going through these sorts of scripts can run ctest and expect it to only try running tests that actually have a chance of passing - requiring a bunch of labels or a script to get a passing run is confusing)

**Review by benvanik — APPROVED** [review]

**Comment by ScottTodd** [comment]
Thinking more about overall infrastructure / CI complexity and I'm not sure if `check` tests are the best tool for the job and are worth further refactoring. May want to replace them with tests authored in Python then converted to `iree-run-module` tests with expected outputs for portable coverage. I could land this PR and continue on a few other refactoring tasks I had started, but now is a good time to take inventory and plan for the year ahead. (Will also find a good place for continued discussions on those topics)

**Review by bjacob — APPROVED** [review]
If you feel that this PR brings the behavior closer to what you would like to be the behavior of the replacement, then it is worth merging it as a stepping stone towards that. Also, unless you are certain that you will have time to do the replacement, there's a chance that we're "stuck with" check tests, so any improvement is good to take.

**Comment by ScottTodd** [comment]
> If you feel that this PR brings the behavior closer to what you would like to be the behavior of the replacement, then it is worth merging it as a stepping stone towards that. Also, unless you are certain that you will have time to do the replacement, there's a chance that we're "stuck with" check tests, so any improvement is good to take.

All good points, thanks for weighing in. Synced and going to merge this.

**Pull request merged**